FLIP: Change the syntax for function types (#43)
onflow/flips: cadence/2022-10-18-change-fun-type-syntax.md

# Change the syntax for function types.

| Status        | (Proposed / Rejected / Accepted / Implemented)       |
:-------------- |:---------------------------------------------------- |
| **FLIP #**    | [43](https://github.com/onflow/flips/pull/43)        |
| **Author(s)** | Naomi Liu (naomi.liu@dapperlabs.com)                 |
| **Updated**   | 2022-10-27                                           |

## Objective

Modify/add syntax for representing function types.

## Motivation

Currently, functions are declared and referred to with differing syntax. For instance,

```cadence
// function declaration
fun foo(_ x: Int): String {...}
// function expression bound to a name
let bar = fun(x: Int): String {...}

// a reference to a function
let baz: ((Int): String) = foo
```

> This convention for representing functions gets even more confusing when currying or higher-order functions are brought into the mix:

Reviewer: Currying is not a common design pattern in Cadence, not worth referring to it here. 

```suggestion
This convention for representing functions gets even more confusing when higher-order functions are brought into the mix:
```

Author: I should open a separate FLIP to allow curried functions like in scala:

```cadence
fun add3(_ x: Int)(_ y: Int)(_ z: Int): Int {
    return x + y + z
}

let plus5 = add3(1)(4)
```

## User Benefit

The existing syntax is inconsistent and unfamiliar, ultimately harming readability of Cadence code. Users would benefit from type signatures with clearer intent.

## Design Proposal

Rename function types to require the `fun` keyword.

Given that `fun` is already a reserved keyword for declaring functions, requiring the keyword in the type signature of a function is more consistent with the existing syntax. For instance, 

    ```cadence
    // (AnyStruct -> Void) -> AnyStruct -> Void
    fun call(f: ((AnyStruct): Void), x: AnyStruct) {...}

    // after
    fun call(f: (fun(AnyStruct): Void), x: AnyStruct) {...}
    ```

This could be implemented as a non-breaking change by still allowing the old syntax of omitting `fun`, but preferring the keyword in new contracts and transactions. The stringification of function types should be updated to use the new syntax.

### Drawbacks

As @turbolent mentioned in an internal discussion, there is some ambiguity in the language grammar regarding the distinction between functions and restricted types. This is what motivated the choice to use the existing syntax without the `fun` keyword.

For example,

```cadence

> fun(): T{}

Reviewer 2: related to my other comment regarding ambiguity: This is not a valid place for a 'type'. i.e: a declaration/expression may not start with a type.
Maybe update it with an example where `fun(): T{}` is placed in a context where a type is expected. I feel there won't be any ambiguity there.

```

Is this a function that returns a `T` but has an empty body? Or does it return a restricted type `T{}` and is missing its body?

One potential route is to give functions a higher priority in the parser than restricted types, so the above expression is parsed as a function returning  `T` with an empty body. Let's take a more complex example:

```cadence
interface Foo {...}

let f = fun(): AnyStruct{Foo}
```

In this case, `f` can be parsed either as a function returning `AnyStruct` whose body only contains the expression `Foo`, or as a function returning a restricted type `AnyStruct{Foo}` but is missing a body. Assume that we prefer to follow the first parsing rule. Since the "body" of the function is missing an explicit `return`, its return type is inferred to be `Void`. However, the stray `Foo` exists in the type namespace instead of the value namespace, so it will result in an unknown name error unless a variable with the same name exists in scope. A lint could be added that ensures statements are either function calls, assignments, declarations, or moves. For instance,

> 

Reviewer 2: This section describes the ambiguity in function expressions. Is it also the case for function types?
If so, maybe adjust this section to describe the ambiguity in function type.

Author: actually wait, this ambiguity only comes into play for expressions. function types themselves don't have a body, so I can probably just cut this section out entirely

Reviewer 2: nice! maybe add it as a remark, saying the ambiguity is not present for function type 👌

```
var x = 0 // declaration, ok
fun f(): Int {log("hi"); return 42} // declaration, ok
x = f() // assignment, ok
f() // function call with unused return, ok
x = x + 1 // ok
x // noop, should throw a lint error
```

Additionally, if the user intends to write a restricted type with multiple interfaces like `AnyStruct{Foo, Bar}`, the parser will throw an error at the use of bare commas outside of a function call or composite literal. It's akin to writing an expression `Foo, Bar`. It seems that the only cases where this ambiguity doesn't throw an error is when the unrestricted return type is parsed as a supertype of `Void`, and the restricting interface is shadowed by a value. 

### Alternatives Considered

> An alternative to using `fun` to denote function types is to introduce another keyword, such as `=>` or `->`. This would be more familiar to users coming from languages such as Javascript or Haskell, while remaining as a non-breaking change due to these operators being unreserved currently. For example, 

Reviewer: To be honest, when I was looking at the syntax for function types in TypeScript, Swift and Kotlin (all of which Cadence explicitly cites as inspiration for its syntax and behavior: https://developers.flow.com/cadence/language/index#syntax-and-behavior), these languages all use `->` for their function types. Given that prepending `fun` to the type as it exists in Cadence today causes a syntax ambiguity, I actually think it might be a better change to use the `->` syntax instead to separate the parameters from the return type. 

I don't think the mismatch between how functions are declared and how their types are expressed is really that big of an issue. Kotlin, for example, uses `->` in its function types despite using declaration syntax that does not use this symbol. 

```suggestion
An alternative to using `fun` to denote function types is to introduce another keyword, such as `=>` or `->`. This would be more familiar to users coming from languages such as JavaScript, Swift, and Kotlin, while remaining as a non-breaking change due to these operators being unreserved currently. For example, 
```

* Does the design conform to the backwards & forwards compatibility [requirements](../docs/compatibility.md)?
* How will this proposal interact with other parts of the Flow Ecosystem?
    - How will it work with FCL?
    - How will it work with the Emulator?
    - How will it work with existing Flow SDKs?

This can be implemented as a non-breaking change, requiring no intervention from developers to update existing code.

### User Impact

* What are the user-facing changes? How will this feature be rolled out?

Function types will be printed using the new syntax in the REPL, error messages, and language server. No migrations or inverventions are needed, and this change can be rolled out with any spork before, during, or after Stable Cadence's release.

## Related Issues

What related issues do you consider out of scope for this proposal,
but could be addressed independently in the future?

If we use a syntax other than `fun` for denoting function types, such as an arrow `->`, the new keyword could be used to implement a shorthand for declaring closures.

## Prior Art

Most languages with first-class functions use similar language for referring to those functions. For instance,

> 

Reviewer 2: Go has a similar syntax to what is proposed  here:
```go
var f func(int) int = ...
```
Would be good to add that too as an example.

# Change the syntax for function types.

| Status        | (Proposed / Rejected / Accepted / Implemented)       |
:-------------- |:---------------------------------------------------- |
| **FLIP #**    | [43](https://github.com/onflow/flips/pull/43)        |
| **Author(s)** | Naomi Liu (naomi.liu@dapperlabs.com)                 |
| **Updated**   | 2022-10-27                                           |

## Objective

Modify/add syntax for representing function types.

## Motivation

Currently, functions are declared and referred to with differing syntax. For instance,

```cadence
// function declaration
fun foo(_ x: Int): String {...}
// function expression bound to a name

> let bar = fun(x: Int): String {...}

Reviewer 3: A space between the keyword and the parameter list is unidiomatic for function expressions. Maybe remove here and for examples of the new function type syntax

```suggestion
let bar = fun (x: Int): String {...}
```

Author: sorry, do you mean idiomatic? as in we suggest authors to insert a space?

// a reference to a function

> let baz: ((Int): String) = foo

Reviewer 3: Maybe explicitly mention that function types have a different syntax from function declarations and function expressions

fun add3(_ x: Int): ((Int): ((Int): Int)) {
    return fun(y: Int): ((Int): Int) {
        return fun(z: Int): Int {
            return x + y + z
        }
    }
}

// cadence type: (([Int], ((Int): String)): [String])
// pretty type: [Int] -> (Int -> String) -> [String]
fun mapIntString(list: [Int], f: ((Int): String)): [String] {...}
```

Another extreme example is the type `(({Int: String}): {Int: Int})`. To a reader who isn't familiar with Cadence, it's not immediately obvious that this type means "a function taking a dictionary of `Int`s to `String`s, returning a dictionary of `Int`s to `Int`s".

## User Benefit

The existing syntax is inconsistent and unfamiliar, ultimately harming readability of Cadence code. Users would benefit from type signatures with clearer intent.

## Design Proposal

Rename function types to require the `fun` keyword.

Given that `fun` is already a reserved keyword for declaring functions, requiring the keyword in the type signature of a function is more consistent with the existing syntax. For instance, 

    ```cadence
    // (AnyStruct -> Void) -> AnyStruct -> Void
    fun apply(f: ((AnyStruct): Void), x: AnyStruct) {...}

    // after

>     fun apply(f: (fun(AnyStruct): Void), x: AnyStruct) {...}

Reviewer 3: For example, here. Also, the parentheses should not be necessary anymore, right?

```suggestion
    fun apply(f: fun (AnyStruct): Void, x: AnyStruct) {...}
```

```cadence
let bar = fun(x: Int): String {...}

// a reference to a function
let baz: ((Int): String) = foo
```

If we continued to elide the `fun` keyword in function types, parentheses could still be eliminated by making `:` a right-associative operator with a higher precedence than `=`. 

```ebnf
func-type = '(' [type {, type}] ')' ':' type
type = func-type | identifier | ...
assignment = identifier [':' type] '=' expression
```

At the same time however, how do we express a function that returns `Void` (`()`) with this syntax? We can omit a return type in function declarations, which causes the type to default to `Void`, but referencing the function requires one to explicitly name `Void`:

```cadence
fun noop() {}

let noop_: ((): Void) = noop
```

Using the `fun` keyword would simplify parsing rules and also allow for return type elision on procedures:

```cadence
let noop_: fun() = noop // unambiguous, since `fun` is our marker now instead of `:`

let baz: fun(Int): String = bar
```

The adjusted grammar would only affect `func-type`:

```ebnf

> func-type = 'fun' '(' [type {, type}] ')' [':' type]

Reviewer 3: This should use the same syntax for rules as the grammar uses in https://github.com/onflow/cadence/blob/master/docs/cadence.ebnf, as well as the same rule names. 

Also, the parameter and result types are not just plain types, but type annotations (`type` -> `typeAnnotation`).

Finally, it looks like this makes the return type optional. Do we want that?

Author: I added a bit of discussion on omitting the return types. I'd argue in favour of it, especially if we go with a syntax like `fun(T) -> U` or `fun(T): U` that's prefixed by a keyword. Languages like Go and Rust keep this omission rule consistent between function declarations and types.

Reviewer 3: Agreed!

let _apply: (AnyStruct -> Void, AnyStruct) -> Void = apply
```

The main benefit of this alternative syntax would be the unambiguous parsing and lower visual noise compared to using `fun`. This would still introduce another inconsistency between how functions are declared and how they're typed, but it's more intuitive at a glance to unfamiliar readers. Introducing function arrows in this way also opens the door to an alternative syntax for denoting closures that doesn't require an annotated return type or `return` statement, which could be useful for simple one-liner functions.

Introducing a separate operator for return types still does not eliminate the previously-mentioned ambiguity in parsing function expressions, however:

```cadence
// still ambiguous, but looks cleaner because ':' in declarations 
// now only indicates a type assignment to a variable
let f: fun() -> AnyStruct{Foo} = fun() -> AnyStruct{Foo} 
```

### Performance Implications

Function types written with either syntax will create identical parse trees, so the performance impact is expected to be limited to another conditional branch in the parser.

### Engineering Impact

> * Do you expect changes to binary size / build time / test times?

Reviewer 3: Remove the points from the template here and in the other sections

# Change the syntax for function types.

> | Status        | (Proposed / Rejected / Accepted / Implemented)       |

Reviewer 3: ```suggestion
| Status        |  Accepted                                            |
```

:-------------- |:---------------------------------------------------- |
| **FLIP #**    | [43](https://github.com/onflow/flips/pull/43)        |
| **Author(s)** | Naomi Liu (naomi.liu@dapperlabs.com)                 |
| **Updated**   | 2022-10-27                                           |

## Objective

Modify/add syntax for representing function types.

## Motivation

Currently, functions are declared and referred to with differing syntax. For instance,

```cadence
// function declaration
fun foo(_ x: Int): String {...}
// function expression bound to a name
let bar = fun (x: Int): String {...}

> // a reference to a function

Reviewer 3: "Reference" is potentially confusing here. Lets only use it when actual references are involved. 
Maybe "a variable declaration with a function value"?

Author: ah good point
